src/ui/public/kuery/lib/ast.js

```diff
@@ -0,0 +1,58 @@
+import grammar from 'raw!./kuery.peg';
+import PEG from 'pegjs';
+import _ from 'lodash';
+
+const kueryParser = PEG.buildParser(grammar);
+
+export function fromKueryExpression(expression) {
+  if (_.isUndefined(expression)) {
+    throw new Error('expression must be a string, got undefined instead');
+  }
+
+  return kueryParser.parse(expression);
+}
+
+export function toKueryExpression(ast) {
+  if (!Array.isArray(ast)) throw new Error('AST must be an array');
+
+  return ast.reduce((acc, node) => {
+    return `${acc} ${stringifyNode(node)}`;
+  }, '').trim();
+}
+
+function stringifyNode(node) {
+  if (!node || !node.type) return '';
+
+  let retVal = '';
+  if (node.negate) retVal += '-';
+
+  switch (node.type) {
```

Comment:
I did a major reorganization in https://github.com/elastic/kibana/pull/11915/commits/d1b37e5a68ef689badde93f7a27efd2ed44d7da5. I'm still not sure I'm 100% happy with it, but I think it's easier to understand than before.